Move generic HashTreeRoot funcs to //shared

Pablosan opened the pull request:
**What type of PR is this?**

Other (enhancement)

**What does this PR do? Why is it needed?**

There are several methods in `stateutils` that would be better suited to a shared library, as they are not directly related to HashTreeRoot states. This PR moves these generic methods to `//shared/htrutils` along with a couple more dependent files.

**Which issues(s) does this PR fix?**

Addresses #5222

**Other notes for review**
  - [X] Move generic methods into a separate `htrutils.go` file, in preparation for moving
  - [x] Move `htrutils.go`, `hash_function.go`, `helpers.go`, and `merkleize.go` from `stateutils` to `//shared/htrutils`
  - [x] Update references to methods in `htrutils.go`, `hash_function.go`, `helpers.go`, and `merkleize.go` to reference their new home
  - [x] Add tests for `htrutils.go`, `hash_function.go`, `helpers.go` and `merkleize.go`

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/prysmaticlabs/prysm?pullRequest=6167) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/prysmaticlabs/prysm?pullRequest=6167) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/prysmaticlabs/prysm?pullRequest=6167) it.</sub>

Pablosan commented:
@nisdas I still need to add tests and run the validator locally before this is ready to be merged, but if you can find the time I wouldn't mind getting your first impression from a quick review of what I have so far. Thanks in advance!

nisdas commented:
@Pablosan On a first look, this all looks good to me :+1: , it seems to be on the right track.

Pablosan commented:
Thanks for the feedback @nisdas! Actually, I think this work might go on for a while so I have a suggestion: can we just review what I have so far in an effort to merge it sooner rather than later? I have removed "WIP" from the title and also changed "Fixes #5222" to "Addresses #5222" so that the Issue doesn't get auto-closed.

That way we can avoid any potentially nasty master merges (either for me or someone else), and I can continue the work in another PR.

Thoughts?

0xKiwi commented on beacon-chain/state/state_trie.go:
```diff
@@ -203,7 +204,7 @@ func (b *BeaconState) HashTreeRoot(ctx context.Context) ([32]byte, error) {
 func merkleize(leaves [][]byte) [][][]byte {
 	hashFunc := hashutil.CustomSHA256Hasher()
-	layers := make([][][]byte, stateutil.GetDepth(uint64(len(leaves)))+1)
+	layers := make([][][]byte, htrutils.GetDepth(uint64(len(leaves)))+1)
```
When do we decide to name `htrutils` vs. `htrutil`? Singular vs. Plural?

Pablosan replied:
@0xKiwi I didn't see a clear pattern in `//shared` so I just picked one. Some of the shared packages are singular, some are plural. Is there a reason to choose one over the other? I'm more than happy to make it `htrutil`. I have no preference.

nisdas replied:
I think `htrutils` is fine here, the plural form makes sense for this.

Pablosan replied:
`//shared/grpcutils` makes it even more confusing. Not only is there only on file, the file only has one function, so ¯\_(ツ)_/¯.

Pablosan commented:
@nisdas I have added tests for all four files moved to `//shared/htrutils` so I think this PR is ready to go. I added another To Do item to revisit `stateutil` and determine what else, if anything, should be moved. I also changed "Fixes" to "Addresses" in reference to the associated issue, which should leave the issue opened so I can continue to reference new PR's to it.

Ready for review!
